### PR TITLE
enhance the scan for kvm

### LINKED ISSUE #895: [FVT] rscan against PowerKVM host cannot detect vmstorage with block device
Below is part of output of command ``rscan``. As you can see, the ``vmstorage`` attribute is empty.

    # rscan c910f03c17 -z
    # <xCAT data object stanza file>

    c910f03c17k22
        arch=ppc64le
        mac=42:05:0a:03:11:16
        mgt=kvm
        vmcpus=4
        vmhost=c910f03c17
        vmmemory=6144
        vmnicnicmodel=virtio
        vmnics=brenP3p9s0f0
        vmstorage=
    c910f03c17k43
        arch=ppc64le
        mac=42:35:0a:03:11:2b
        mgt=kvm
        vmcpus=4
        vmhost=c910f03c17
        vmmemory=6144
        vmnicnicmodel=virtio
        vmnics=brenP3p9s0f0
        vmstorage=

Actually, these PowerKVM guests use block devices as the ``vmstorage``s.

    # lsdef c910f03c17k22
    Object name: c910f03c17k22
        arch=ppc64el
        currchain=boot
        currstate=boot
        groups=all
        id=22
        initrd=xcat/osimage/ubuntu14.04.4-ppc64el-install-compute/initrd.img
        kcmdline=nofb utf8 auto url=http://!myipfn!/install/autoinst/c910f03c17k22 xcatd=!myipfn! mirror/http/hostname=!myipfn! netcfg/choose_interface=42:05:0a:03:11:16 console=tty0 console=hvc0,115200 locale=en_US priority=critical hostname=c910f03c17k22 live-installer/net-image=http://!myipfn!/install/ubuntu14.04.4/ppc64el/install/filesystem.squashfs
        kernel=xcat/osimage/ubuntu14.04.4-ppc64el-install-compute/vmlinuz
        mac=42:05:0a:03:11:16|42:03:0a:03:11:16!*NOIP*
        mgt=kvm
        netboot=grub2
        os=ubuntu14.04.4
        postbootscripts=otherpkgs
        postscripts=syslog,remoteshell,syncfiles
        profile=compute
        provmethod=ubuntu14.04.4-ppc64el-install-compute
        serialport=0
        serialspeed=115200
        status=booted
        statustime=03-22-2016 05:02:37
        vmcpus=4
        vmhost=c910f03c17
        vmmemory=6144
        vmnics=brenP3p9s0f0,brenP3p9s0f1
        vmstorage=phy:/dev/mapper/vdiskvg03-vdisk01n22

### LINKED ISSUE #899: [FVT] rscan against PowerKVM can only detect the first vmnics
Below is part of output of command ``rscan``. As you can see, only the first ``vmnics`` can be detected.

    # rscan c910f03c17 -z
    # <xCAT data object stanza file>

    c910f03c17k22
        arch=ppc64le
        mac=42:05:0a:03:11:16
        mgt=kvm
        vmcpus=4
        vmhost=c910f03c17
        vmmemory=6144
        vmnicnicmodel=virtio
        vmnics=brenP3p9s0f0
        vmstorage=
    c910f03c17k43
        arch=ppc64le
        mac=42:35:0a:03:11:2b
        mgt=kvm
        vmcpus=4
        vmhost=c910f03c17
        vmmemory=6144
        vmnicnicmodel=virtio
        vmnics=brenP3p9s0f0
        vmstorage=

Actually, these PowerKVM guests have more than one ``vmnics``.

    # lsdef c910f03c17k22
    Object name: c910f03c17k22
        arch=ppc64el
        currchain=boot
        currstate=boot
        groups=all
        id=22
        initrd=xcat/osimage/ubuntu14.04.4-ppc64el-install-compute/initrd.img
        kcmdline=nofb utf8 auto url=http://!myipfn!/install/autoinst/c910f03c17k22 xcatd=!myipfn! mirror/http/hostname=!myipfn! netcfg/choose_interface=42:05:0a:03:11:16 console=tty0 console=hvc0,115200 locale=en_US priority=critical hostname=c910f03c17k22 live-installer/net-image=http://!myipfn!/install/ubuntu14.04.4/ppc64el/install/filesystem.squashfs
        kernel=xcat/osimage/ubuntu14.04.4-ppc64el-install-compute/vmlinuz
        mac=42:05:0a:03:11:16|42:03:0a:03:11:16!*NOIP*
        mgt=kvm
        netboot=grub2
        os=ubuntu14.04.4
        postbootscripts=otherpkgs
        postscripts=syslog,remoteshell,syncfiles
        profile=compute
        provmethod=ubuntu14.04.4-ppc64el-install-compute
        serialport=0
        serialspeed=115200
        status=booted
        statustime=03-22-2016 05:02:37
        vmcpus=4
        vmhost=c910f03c17
        vmmemory=6144
        vmnics=brenP3p9s0f0,brenP3p9s0f1
        vmstorage=phy:/dev/mapper/vdiskvg03-vdisk01n22

### PR DESCRIPTION
**Pull request by penguhyang**
fix  #899
fix #895 
Document link:
http://10.2.9.4/guides/admin-guides/references/man1/rscan.1.html?highlight=rscan

### CODE REVIEW

**Comment by immarvin** [comment]
need to reconsider the behavior of rscan for vm

File: xCAT-server/lib/xcat/plugins/kvm.pm [hunk 1]
```diff
@@ -2885,6 +2885,12 @@ sub mkvm {
         "cpus=s"     => \$cpucount,
         'force|f'    => \$force
     );
+    if (defined $confdata->{vm}->{$node}->[0]->{othersettings}) {
+        my $nodefrom = $confdata->{vm}->{$node}->[0]->{othersettings};
+        if ($nodefrom =~ /^nodefromrscan/) {
```
Comment (immarvin) [inline xCAT-server/lib/xcat/plugins/kvm.pm]: ``$nodefrom =~ /nodefromrscan/``? since, "nodefromrscan" might not be the beginning of "othersettings"

Comment (penguhyang) [inline xCAT-server/lib/xcat/plugins/kvm.pm]: fixed.

**Comment by immarvin** [comment]
"rscan -z" should output the "vm.othersettings" attribute

**Comment by penguhyang** [comment]
After discussion in China Team.
For KVM host, we will not support -z option for rscan at present time.

**Comment by immarvin** [comment]
please attach the link of the updated documentation

File: xCAT-server/lib/xcat/plugins/kvm.pm [hunk 2]
```diff
@@ -2885,6 +2885,12 @@ sub mkvm {
         "cpus=s"     => \$cpucount,
         'force|f'    => \$force
     );
+    if (defined $confdata->{vm}->{$node}->[0]->{othersettings}) {
+        my $nodefrom = $confdata->{vm}->{$node}->[0]->{othersettings};
+        if ($nodefrom =~ /nodefromrscan/) {
+            return 1, "this node was defined through rscan, it does not support to use 'mkvm' to create again.";
```
Comment (immarvin) [inline xCAT-server/lib/xcat/plugins/kvm.pm]: "mkdef is not supported"

File: xCAT-server/lib/xcat/plugins/kvm.pm [hunk 3]
```diff
@@ -2885,6 +2885,12 @@ sub mkvm {
         "cpus=s"     => \$cpucount,
         'force|f'    => \$force
     );
+    if (defined $confdata->{vm}->{$node}->[0]->{othersettings}) {
+        my $nodefrom = $confdata->{vm}->{$node}->[0]->{othersettings};
```
Comment (immarvin) [inline xCAT-server/lib/xcat/plugins/kvm.pm]: the "$nodefrom" is not a valid variable name, suggest it to be "$vmothersettings"

File: xCAT-server/lib/xcat/plugins/kvm.pm (hunk 2 above)
Comment (immarvin) [inline xCAT-server/lib/xcat/plugins/kvm.pm]: should be ``return (1, "this node was defined through rscan, it does not support to use 'mkvm' to create again.");``

File: xCAT-server/lib/xcat/plugins/kvm.pm [hunk 4]
```diff
@@ -3136,14 +3138,23 @@ sub rscan {
...
                 if (@vmstorageobj and defined($vmstorageobj[0])) {
-                    $vmstorage = $vmstorageobj[0]->getAttribute("file");
-                    last;
+                    $vmstorage_file_obj = $vmstorageobj[0]->getAttribute("file");
                 }
+                $vmstorage .= "$vmstorage_file_obj,";
+            }
+            if (($vmstoragediskobj->getAttribute("device") eq "disk") and ($vmstoragediskobj->getAttribute("type") eq "block")) {
+                my @vmstorageobj = $vmstoragediskobj->findnodes("./source");
+                if (@vmstorageobj and defined($vmstorageobj[0])) {
```
Comment (immarvin) [inline xCAT-server/lib/xcat/plugins/kvm.pm]: ``if (@vmstorageobj)`` is enough

File: xCAT-server/lib/xcat/plugins/kvm.pm (hunk 3 above)
Comment (penguhyang) [inline xCAT-server/lib/xcat/plugins/kvm.pm]: fixed.

File: xCAT-server/lib/xcat/plugins/kvm.pm (hunk 2 above)
Comment (penguhyang) [inline xCAT-server/lib/xcat/plugins/kvm.pm]: fixed.

File: xCAT-server/lib/xcat/plugins/kvm.pm (hunk 4 above)
Comment (penguhyang) [inline xCAT-server/lib/xcat/plugins/kvm.pm]: fixed.

**Comment by immarvin** [comment]
seems ok to me, agree to merge

**Pull request merged**